### PR TITLE
#77: refactor list_exists rule

### PR DESCRIPTION
**Pull request by yogyrton**
https://github.com/RonasIT/laravel-helpers/issues/77

### CODE REVIEW

File: src/HelpersServiceProvider.php [hunk 1]
```diff
@@ -76,12 +75,20 @@ protected function extendValidator(): void
             $keyField = Arr::get($parameters, 1, 'id');
 
             if (!empty(Arr::get($parameters, 2))) {
-                $value = collect($value)->pluck(Arr::get($parameters, 2));
+                $value = collect($value)
+                    ->pluck(Arr::get($parameters, 2))
+                    ->toArray();
```
Comment (aizlee) [inline src/HelpersServiceProvider.php:80]:
```suggestion
                $value = Arr::pluck($value, Arr::get($parameters, 2));
```

File: src/HelpersServiceProvider.php [hunk 2]
```diff
@@ -76,12 +75,20 @@ protected function extendValidator(): void
...
             }
 
-            return DB::table($table)
+            $value = array_unique($value);
+
+            $existingValue = DB::table($table)
```
Comment (aizlee) [inline src/HelpersServiceProvider.php:85]:
```suggestion
            $existingValueCount = DB::table($table)
```

File: src/HelpersServiceProvider.php [hunk 3]
```diff
@@ -76,12 +75,20 @@ protected function extendValidator(): void
...
                 ->whereIn($keyField, $value)
-                ->exists();
+                ->distinct()
+                ->pluck($keyField)
+                ->toArray();
```
Comment (aizlee) [inline src/HelpersServiceProvider.php:89]:
```suggestion
                ->distinct()
                ->count();
```

**Review by aizlee — CHANGES_REQUESTED** [review]

File: src/HelpersServiceProvider.php (hunk 3 above)
Comment (yogyrton) [reply src/HelpersServiceProvider.php:89]: We have to use `pluck($keyField)` before `distinct()` because without it all the entries will be unique

Comment (aizlee) [reply src/HelpersServiceProvider.php]: You are right. 

We can modify the code to:
```
DB::table($table)
                ->whereIn($keyField, $value)
                ->distinct()
                ->count($keyField);
```

In this case `distinct` will work correctly with `count`

**Comment by aizlee** [comment]
@yogyrton We need to check if we provide duplicate values the `list_exists`  method works correctly. Please add a test for this case.

**Review by aizlee — APPROVED** [review]

File: tests/support/Traits/SqlMockTrait.php [hunk 4]
```diff
@@ -432,14 +432,16 @@ protected function mockExistsUsersExceptAuthorizedByArray(
...
-        $this->mockSelectExists(
-            "select exists(select * from \"{$table}\" where \"{$keyField}\" in (?, ?, ?)) as \"exists\"",
-            $isExist,
-            [1, 2, 3]
+        $countResult = [[ 'aggregate' => count($result) ]];
+
+        $this->mockSelect(
+            query: "select count(distinct \"{$keyField}\") as aggregate from \"{$table}\" where \"{$keyField}\" in (?, ?, ?)",
+            result: $countResult,
+            bindings: [1, 2, 3],
```
Comment (DenTray) [inline tests/support/Traits/SqlMockTrait.php:444]:
```suggestion
            bindings: $result,
```

File: tests/support/Traits/SqlMockTrait.php [hunk 5]
```diff
@@ -432,14 +432,16 @@ protected function mockExistsUsersExceptAuthorizedByArray(
...
     ): void {
-        $this->mockSelectExists(
-            "select exists(select * from \"{$table}\" where \"{$keyField}\" in (?, ?, ?)) as \"exists\"",
-            $isExist,
-            [1, 2, 3]
+        $countResult = [[ 'aggregate' => count($result) ]];
+
+        $this->mockSelect(
+            query: "select count(distinct \"{$keyField}\") as aggregate from \"{$table}\" where \"{$keyField}\" in (?, ?, ?)",
+            result: $countResult,
```
Comment (DenTray) [inline tests/support/Traits/SqlMockTrait.php:443]:
```suggestion
        $this->mockSelect(
            query: "select count(distinct \"{$keyField}\") as aggregate from \"{$table}\" where \"{$keyField}\" in (?, ?, ?)",
            result: [[ 'aggregate' => count($result) ]],
```

File: tests/ValidatorTest.php [hunk 6]
```diff
@@ -77,9 +77,21 @@ public function testListExists()
...
     }
 
+    public function testListExistsIfDuplicateValues()
+    {
+        $this->mockListExists([1, 2, 3]);
+
+        $validator = Validator::make(
+            ['ids' => [1, 2, 3, 3]],
+            ['ids' => 'list_exists:clients,user_id'],
+        );
```
Comment (DenTray) [inline tests/ValidatorTest.php:87]:
```suggestion
        $validator = Validator::make([
            'ids' => [1, 2, 3, 3],
        ] [
            'ids' => 'list_exists:clients,user_id',
        ]);
```

**Review by DenTray — APPROVED** [review]

File: tests/ValidatorTest.php [hunk 7]
```diff
@@ -77,9 +77,22 @@ public function testListExists()
         $this->assertTrue($validator->passes());
     }
 
+    public function testListExistsIfDuplicateValues()
+    {
+        $this->mockListExists([1, 2, 3]);
+
+        $validator = Validator::make([
+            'ids' => [1, 2, 3, 3],
+        ] [
```
Comment (DenTray) [inline tests/ValidatorTest.php:86]:
```suggestion
        ], [
```

**Review by DenTray — APPROVED** [review]

**Comment by DenTray** [comment]
@yogyrton please check failed tests, it fails after https://github.com/RonasIT/laravel-helpers/pull/158#discussion_r1881244074

**Review by DenTray — APPROVED** [review]

**Pull request merged**